File: backend/app/schemas/reserva.py
from datetime import date, datetime, time
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


ReservaEstado = Literal["esperando", "aprobada", "rechazada", "cancelada"]


class ReservaCreate(BaseModel):
    espacio_id: int
    fecha: date
    hora_inicio: time
    hora_fin: time
    asistentes: int = Field(gt=0)


class ReservaEstadoUpdate(BaseModel):
    nuevo_estado: Literal["aprobada", "rechazada"]


class UsuarioReservaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    username: str
    email: str
    rol: str


class EspacioReservaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    nombre: str
    capacidad: int
    estado: str


class ReservaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    usuario_id: int
    espacio_id: int
    fecha: date
    hora_inicio: time
    hora_fin: time
    estado: ReservaEstado
    asistentes: int
    created_at: datetime
    updated_at: datetime
    usuario: UsuarioReservaResponse
    espacio: EspacioReservaResponse

    @field_validator("estado")
    @classmethod
    def validar_estado(cls, value: str) -> str:
        if value not in {"esperando", "aprobada", "rechazada", "cancelada"}:
            raise ValueError("Estado de reserva inválido")
        return value

File: frontend/src/types/reserva.ts
export type ReservaEstado = 'esperando' | 'aprobada' | 'rechazada' | 'cancelada';

export interface ReservaUsuario {
  id: number;
  username: string;
  email: string;
  rol: 'admin' | 'usuario' | string;
}

export interface ReservaEspacio {
  id: number;
  nombre: string;
  capacidad: number;
  estado: 'activo' | 'inactivo' | 'mantenimiento' | string;
}

export interface Reserva {
  id: number;
  usuario_id: number;
  espacio_id: number;
  fecha: string;
  hora_inicio: string;
  hora_fin: string;
  estado: ReservaEstado;
  asistentes: number;
  created_at: string;
  updated_at: string;
  usuario: ReservaUsuario;
  espacio: ReservaEspacio;
}

export interface ReservaCreate {
  espacio_id: number;
  fecha: string;
  hora_inicio: string;
  hora_fin: string;
  asistentes: number;
}

export interface ReservaEstadoUpdate {
  nuevo_estado: Extract<ReservaEstado, 'aprobada' | 'rechazada'>;
}
